### PR TITLE
fix(macOS/#1085): allow opening of files through system protocol

### PR DESCRIPTION
**Pull request by zbaylin**
As of right now, if you try to open a file in Oni on macOS either through "Open With" in Finder or with the `open` command in Terminal.app, Oni activates but doesn't open the file passed in.

This does two main things:
- adds file associations to the `Info.plist` so that we're shown when trying to "Open With" a compatible file:
image

- binds to Revery's new `App.onFileOpen` event system to trigger opens within Oni2 itself

It also changes the executable in the `Info.plist` from `Oni2` to `Oni2_editor` -- this is necessary because otherwise the file information gets lost between the processes. From my conversations with @bryphe, this seems to be okay, I just have to make sure that Gatekeeper doesn't interfere with CLI flags.

Addresses #1085. Thank you @raguay for bringing this to our attention!

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -18,6 +18,70 @@
...
+        ]
+      },
+      {
+        "name": "ReasonML source",
+        "role": "Editor",
+        "ext": [
+          ".re",
+          ".rei"
+        ]
+      },
```
Comment (bryphe) [inline package.json:76]: Thanks for adding these additional file associations!

File: scripts/release.js [hunk 2]
```diff
@@ -123,8 +123,15 @@ if (process.platform == "linux") {
...
-      CFBundleExecutable: "Oni2",
+      CFBundleExecutable: "Oni2_editor",
       NSHighResolutionCapable: true,
+      CFBundleDocumentTypes: package.build.fileAssociations.map(fileAssoc => {
+            return {
+                CFBundleTypeExtensions: fileAssoc.ext.map(ext => ext.substr(1)),
+                CFBundleTypeName: fileAssoc.name,
+                CFBundleTypeRole: fileAssoc.role
+            }
+        })
```
Comment (bryphe) [inline scripts/release.js:134]: nice! 👍

File: scripts/release.js [hunk 3]
```diff
@@ -123,8 +123,15 @@ if (process.platform == "linux") {
       CFBundleVersion: `${package.version}`,
       CFBundlePackageType: "APPL",
       CFBundleSignature: "????",
-      CFBundleExecutable: "Oni2",
+      CFBundleExecutable: "Oni2_editor",
```
Comment (bryphe) [inline scripts/release.js:126]: This looks good, but we'll need to port over this CLI logic:
https://github.com/onivim/oni2/blob/8c600e062746eb647dd8270cbc770de2365f51dc/src/bin_launcher/Oni2.re#L97

Comment (zbaylin) [reply scripts/release.js:126]: I noted that in the original PR comment 👍. I was looking at `Cli.re`, but it seems to me that we dont pass in `Sys.argv` like we do in `Oni2`. Not sure what the best solution for that is, unfortunately.

**Comment by bryphe** [comment]
Regarding the command-line-flag that needs to be ported over, it can be a bit tricky to test! That's because, as far as I know, there isn't a very reliable way to set the 'gatekeeper' quarantine flag, reliably.

So the steps I use to test this are:

Setup: Install `npm install -g http-server` to have it around

1) Create the `_release/Onivim2.app` folder via `esy create-release`
2) Run `./scripts/osx/package.sh` to create a `tar.gz` in the `publish` folder
3) `cd _publish`
4) `http-server .` <-- start a local http server in the `publish` folder
5) Go to a web browser, and navigate to `http://127.0.0.1:8080/Onivim2-<commitid>-tar.gz`
6) The tar.gz should be downloaded
7) Extract the `tar.gz`, and run Onivim 2. When you download a file - even if its locally - the gatekeeper quarantine bit gets set!
8) Because the app isn't code signed, you'll need to go to `Security & Privacy` and allow it to open
9) Try opening again - this time, you'll get a warning, but you'll be able to Open

__Expected:__ Application should start successfully
__Actual:__ You can see icon bounce in the dock, but it fails to start. Starting it again works!

The reason it's failing to start is because its being passed that extra `-psn` parameter that isn't handled. Some details on when we hit this before: https://github.com/onivim/oni2/issues/552

File: scripts/release.js (hunk 3 above)
Comment (bryphe) [reply scripts/release.js:126]: Ah, ya, we need to use a different call - `Arg.argv_parse` - that lets us pass in a list of arguments (`Argv.parse` is like `Arg.argv_parse` with `Sys.argv` passed in). 

I made the change here: https://github.com/onivim/oni2/pull/1795 (and wanted to add some test coverage on our CLI anyway, so this was a good opportunity)

**Comment by bryphe** [comment]
Looks like this needs an `esy format`

**Comment by zbaylin** [comment]
@bryphe I somehow missed that! Thanks! Will fix now.

**Pull request merged**